### PR TITLE
Bug fix/optimizations 18022019

### PR DESCRIPTION
**Pull request by jsteemann**
A bunch of micro-optimizations

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3053/

File: arangod/Aql/AqlValue.cpp [hunk 1]
```diff
@@ -552,6 +553,50 @@ AqlValue AqlValue::get(CollectionNameResolver const& resolver,
...
+    // intentionally falls through
+    case VPACK_INLINE:
+    // intentionally falls through
+    case VPACK_MANAGED_SLICE:
+    // intentionally falls through
+    case VPACK_MANAGED_BUFFER: {
+      VPackSlice s(slice());
+      if (s.isObject()) {
+        VPackSlice found(s.get(name));
+        if (found.isCustom()) {
```
Comment (graetzer) [inline arangod/Aql/AqlValue.cpp]: do we actually validate that the only custom field is `_id` ?

File: arangod/Aql/AqlValue.cpp [hunk 2]
```diff
@@ -552,6 +553,50 @@ AqlValue AqlValue::get(CollectionNameResolver const& resolver,
   return AqlValue(AqlValueHintNull());
 }
 
+/// @brief get the (object) element by name
+AqlValue AqlValue::get(CollectionNameResolver const& resolver,
+                       arangodb::velocypack::StringRef const& name, bool& mustDestroy, bool doCopy) const {
```
Comment (graetzer) [inline arangod/Aql/AqlValue.cpp]: can we perhaps typedef the `lib/Basics/StringRef.h` class to the velocypack one ?

File: arangod/Aql/ExecutionNode.cpp [hunk 3]
```diff
@@ -111,9 +111,15 @@ std::unordered_map<int, std::string const> const typeNames{
 // during register planning
 bool isInSubQuery(ExecutionNode const* node) {
   auto current = node;
-  while (current->hasDependency()){
+  TRI_ASSERT(current != nullptr);
+  while (current != nullptr && current->hasDependency()) {
     current = current->getFirstDependency();
   }
+  if (current == nullptr) {
```
Comment (graetzer) [inline arangod/Aql/ExecutionNode.cpp]:
```suggestion
  if (ADB_UNLIKELY(current == nullptr)) {
```

File: arangod/RocksDBEngine/RocksDBVPackIndex.h [hunk 4]
```diff
@@ -126,7 +124,25 @@ class RocksDBVPackIndexIterator final : public IndexIterator {
   bool hasCovering() const override { return true; }
 
  private:
-  bool outOfRange() const;
+  inline bool outOfRange() const {
+    // TODO: prove that we can remove this extra check completely and
+    // rely on iterate_upper/lower_bound
+    if (_reverse) {
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBVPackIndex.h]: we cannot ATM

File: arangod/Aql/AqlValue.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/AqlValue.cpp]: At the moment, we don't validate this. But all our code uses the custom field type just for `_id`.

File: arangod/RocksDBEngine/RocksDBVPackIndex.h (hunk 4 above)
Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBVPackIndex.h]: ok, I am removing the comment...

File: arangod/Aql/AqlValue.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/AqlValue.cpp]: good idea. will do that!

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3054/

**Pull request merged**